crates/hyperdrive-math/src/lp.rs

```diff
@@ -353,6 +412,164 @@ impl State {
+
+    /// Note that the state is the present state of the pool and original values
+    /// passed in as parameters.  Present sate variables are not expressly
+    /// paased in because so that downstream function like kUp() can still be
+    /// used.
+    ///
+    /// Given a signed bond amount, this function calculates the negation
+    /// of the derivative of `calculateSharesOutGivenBondsIn` when the
+    /// bond amount is positive or the derivative of
+    /// `calculateSharesInGivenBondsOut` when the bond amount is negative.
+    /// In both cases, the calculation is given by:
+    ///
+    ///  derivative = (1 - zeta / z) * (
+    ///      1 - (1 / c) * (
+    ///          c * (mu * z_e(x)) ** -t_s +
+    ///          (y / z_e) * y(x) ** -t_s  -
+    ///          (y / z_e) * (y(x) + dy) ** -t_s
+    ///      ) * (
+    ///          (mu / c) * (k(x) - (y(x) + dy) ** (1 - t_s))
+    ///      ) ** (t_s / (1 - t_s))
+    ///  )
+    ///
+    ///  This quantity is used in Newton's method to search for the optimal
+    ///  share proceeds. When the pool is net long, We can express the
+    ///  derivative of the objective function F(x) by the derivative
+    ///  -z_out'(x) that this function returns:
+    ///
+    ///  -F'(x) = l * -PV'(x)
+    ///         = l * (1 - net_c'(x))
+    ///         = l * (1 + z_out'(x))
+    ///         = l * (1 - derivative)
+    ///
+    ///  When the pool is net short, we can express the derivative of the
+    ///  objective function F(x) by the derivative z_in'(x) that this
+    ///  function returns:
+    ///
+    ///  -F'(x) = l * -PV'(x)
+    ///         = l * (1 - net_c'(x))
+    ///         = l * (1 - z_in'(x))
+    ///         = l * (1 - derivative)
+    ///
+    ///  With these calculations in mind, this function rounds its result
+    ///  down so that F'(x) is overestimated. Since F'(x) is in the
+    ///  denominator of Newton's method, overestimating F'(x) helps to avoid
+    ///  overshooting the optimal solution.
+    fn calculate_shares_delta_given_bonds_delta_derivative(
+        &self,
+        bond_amount: I256,
+        original_share_reserves: FixedPoint,
+        original_bond_reserves: FixedPoint,
+        original_effective_share_reserves: FixedPoint,
+        original_share_adjustment: I256,
+    ) -> Result<FixedPoint> {
+        // Calculate the bond reserves after the bond amount is applied.
+        let bond_reserves_after = if bond_amount >= I256::zero() {
+            self.bond_reserves() + bond_amount.into()
+        } else {
+            let bond_amount = FixedPoint::from(U256::try_from(-bond_amount)?);
+            if bond_amount < self.bond_reserves() {
+                self.bond_reserves() - bond_amount
+            } else {
+                return Err(eyre!("Calculating the bond reserves underflows"));
+            }
+        };
+
+        // NOTE: Round up since this is on the rhs of the final subtraction.
+        //
+        // derivative = c * (mu * z_e(x)) ** -t_s +
```

Comment:
This uses `-t_s` but the code uses `self.time_stretch()`. In solidity we have to map `t_s` to `-t_s` but I think we do not in rust. Honestly I get confused by this so we may want to double check with @jalextowle , but either way I think the comment should match the code in Rust.

Reply:
Also the code does `c / (...)` but the comment says `c * (...)`

Reply:
it does match, the -1 part of the exponent is dropped and the base is flipped.  the code is altered from the comment to deal with lack of fixedpoints negative number capability.

Reply:
Ahh so it's positive in the denominator; got it. Can you add a comment mentioning this? I'm trying to make sure that comments match code as much as possible, or have explanations whenever it doesn't.

Reply:
also fyi there's a general problem that the comment math doesn't always match what is implemented. It's residual from new people adding comments as part of their onboarding and doing it incorrectly :(